File: merge_sort/mergeSort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

void merge(int* arr, int lo, int hi, int mid)
{

    int* temp = (int*)malloc((hi-lo+1) * sizeof(int));

    if (arr == NULL) 
    {
        perror("Memory allocation failed");
        exit(1); 
    }

    int i, j, k;
    i = lo;
    k = 0;
    j = mid + 1;
 
    while (i <= mid && j <= hi)
    {
        if (arr[i] < arr[j])
        {
            temp[k] = arr[i];
            k++;
            i++;
        }
        else
        {
            temp[k] = arr[j];
            k++;
            j++;
        }
    }

    while (i <= mid)
    {
        temp[k] = arr[i];
        k++;
        i++;
    }
 
    while (j <= hi)
    {
        temp[k] = arr[j];
        k++;
        j++;
    }
 
    for (i = lo; i <= hi; i++)
    {
        arr[i] = temp[i-lo];
    }

    free(temp);
}


void mergeSort(int* arr, int lo, int hi) 
{

    int mid;

    if (lo < hi) 
    {
        mid = (lo + hi) / 2;

        mergeSort(arr, lo, mid);
        mergeSort(arr, mid+1, hi);
 
        merge(arr, lo, hi, mid);
    }
}

int* generateArray(int n) 
{

    int* arr = (int*)malloc(n * sizeof(int)); 

    if (arr == NULL) 
    {
        perror("Memory allocation failed");
        exit(1); 
    }

    for (int i = 0; i < n; i++) 
    {
        arr[i] = i + 1;
    }

    return arr;
}

int* generateRandomArray(int n) 
{

    int* arr = generateArray(n);

    srand(time(NULL));

    for (int i = n - 1; i > 0; i--) 
    {
        int j = rand() % (i + 1);
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    return arr;
}

bool compareArrays(int arr1[], int arr2[], int size) 
{
    for (int i = 0; i < size; i++) 
    {
        if (arr1[i] != arr2[i]) {
            return false; 
        }
    }
    return true; 
}

int main(int argc, char *argv[]) {
    
    int n = atoi(argv[1]);

    int* correctArray = generateArray(n);

    int* randomArray = generateRandomArray(n);

    clock_t start_time, end_time;
    double cpu_time_used;

    start_time = clock();
    mergeSort(randomArray, 0, n-1);
    end_time = clock();
    cpu_time_used = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

    bool isCorrect = compareArrays(randomArray, correctArray, n);

    free(correctArray);
    free(randomArray);

    if (isCorrect) {
        printf("Correct, boi\n");
    } else {
        printf("Wrong, boi\n");
    }

    printf("Execution Time: %f seconds\n", cpu_time_used);
}

File: PY/merge_sort.py
import random
import time
import sys

def merge(arr, l, m, r):
    n1 = m - l + 1
    n2 = r - m
 
    L = [0] * (n1)
    R = [0] * (n2)
 
    for i in range(0, n1):
        L[i] = arr[l + i]
 
    for j in range(0, n2):
        R[j] = arr[m + 1 + j]
 
    i = 0   
    j = 0     
    k = l     
 
    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1
 
    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1
 
    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1
 
def merge_sort(arr, l, r):
    if l < r:
 
        m = l+(r-l)//2
 
        merge_sort(arr, l, m)
        merge_sort(arr, m+1, r)
        merge(arr, l, m, r)
 


def generate_random_array(length):
    array = list(range(1, length + 1))  
    random.shuffle(array)  
    return array

def compare_arrays(array1, array2):
    if len(array1) != len(array2):
        return False

    for i in range(len(array1)):
        if array1[i] != array2[i]:
            return False

    return True

def test(n):

    n = int(n);    

    random_array = generate_random_array(n)

    correct_array = list(range(1, n + 1)) 

    start_time = time.time()
    merge_sort(random_array, 0, n - 1)
    end_time = time.time()
    execution_time = end_time - start_time


    print(compare_arrays(random_array, correct_array)) 
    print(f"Function execution time: {execution_time:.6f} seconds")

if len(sys.argv) < 2:
    print("Usage: python script.py <argument>")
    sys.exit(1)

script_name = sys.argv[0]
argument = sys.argv[1]

test(argument)

print("Script name:", script_name)
print("Argument:", argument)